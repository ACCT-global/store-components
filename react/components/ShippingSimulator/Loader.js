import React from 'react'
import ContentLoader from 'react-content-loader'

import shippingSimulator from './shippingSimulator.css'

const ShippingSimulatorLoader = (loaderProps = {}) => (
  <div className={`${shippingSimulator.shippingContainer}`}>
    <ContentLoader
      className={`${shippingSimulator.shippingContainerLoader}`}
      style={{
        width: '100%',
        height: '100%',
      }}
      width={500}
      height={40}
      preserveAspectRatio="xMinYMin meet"
      {...loaderProps}
    >
      <rect
        height="100%"
        width="7em"
        {...loaderProps[`${shippingSimulator.shippingZipcodeLabelLoader}`]}
      />
      <rect
        height="100%"
        width="15em"
        x="8em"
        {...loaderProps[`${shippingSimulator.shippingInputLoader}`]}
      />
    </ContentLoader>
  </div>
)

export default ShippingSimulatorLoader
