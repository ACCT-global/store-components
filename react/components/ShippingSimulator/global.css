.vtex-shipping-simulator {
  height: 1.85em;
}

.vtex-shipping-simulator .vtex-input {
  height: 100%;
  margin-left: 0.625em;
}

.vtex-shipping-simulator .vtex-input input {
  padding: 5px 5px;
  color: #4f4f4f;
  border: 1px solid #bdbdbd;
  border-radius: 2px;
  height: 100%;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.vtex-shipping-simulator .vtex-button {
  height: 100%;
  width: 26px;
  font-size: inherit;
  background-color: #e0e0e0;
  padding: 6px 2px;
  border: 1px solid #bdbdbd;
  border-left: none;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}
