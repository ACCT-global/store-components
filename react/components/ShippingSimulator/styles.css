.shippingContainer {}
.shippingContainerLoader {}
.shippingZipcodeLabelLoader {}
.shippingInputLoader {}
.shippingZipcodeLabel {}
.shippingCTA {}
.shippingNoMessage {}
.shippingTable {}
.shippingTableCell {}
.shippingTableLabel {}
.shippingTableRadioBtn {}

.shippingContainer :global(.vtex-input) {
  height: 100%;
}

.shippingContainer :global(.vtex-input input) {
  padding: 10px 5px;
  color: #4f4f4f;
  border: 1px solid #e3e4e6;
  border-radius: 2px;
  height: 100%;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.shippingContainer :global(.vtex-button) {
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}