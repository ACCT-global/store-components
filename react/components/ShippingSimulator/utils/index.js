let addressId = 1

const getRandomAddressId = () => {
  return (addressId++ * new Date().getTime() * -1).toString().replace('-', '')
}

export const getNewAddress = country => {
  const addressId = getRandomAddressId()
  return {
    addressId: addressId,
    addressType: 'residential',
    city: null,
    complement: null,
    country: country,
    geoCoordinates: [],
    neighborhood: null,
    number: null,
    postalCode: null,
    receiverName: null,
    reference: null,
    state: null,
    street: null,
    addressQuery: null,
  }
}
