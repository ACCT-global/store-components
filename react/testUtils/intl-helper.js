import React from 'react'
import * as reactTestingLibrary from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defaultStrings from '../../messages/en-US.json'
const Context = React.createContext({})

export const render = (node, options) => {
  const rendered = reactTestingLibrary.render(
    <IntlProvider messages={defaultStrings} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      render(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}

// re-export everything
export default reactTestingLibrary
