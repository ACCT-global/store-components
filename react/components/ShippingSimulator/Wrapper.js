import React, { useContext } from 'react'
import { ProductContext } from 'vtex.product-context'
import { path, isEmpty } from 'ramda'
import { useRuntime } from 'vtex.render-runtime'

import ShippingSimulator from './index'

const ShippingSimulatorWrapper = (props) => {
  const { culture } = useRuntime()
  const valuesFromContext = useContext(ProductContext)

  const shippingSimultatorProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) {
      return props
    }

    const { selectedItem } = valuesFromContext

    return {
      ...props,
      skuId: path(['itemId'], selectedItem),
      seller: path(['sellers', 0, 'sellerId'], selectedItem),
      country: culture.country,
    }
  }

  return (
    <ShippingSimulator { ...shippingSimultatorProps() } />
  )
}

export default ShippingSimulatorWrapper