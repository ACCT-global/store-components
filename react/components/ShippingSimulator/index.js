import React, { Fragment, useState } from 'react'
import { injectIntl, intlShape } from 'react-intl'
import { compose, withApollo } from 'react-apollo'
import PropTypes from 'prop-types'
import { Button } from 'vtex.styleguide'
import { StyleguideInput } from 'vtex.address-form/inputs/'
import {
  AddressRules,
  AddressContainer,
  PostalCodeGetter,
} from 'vtex.address-form'
import { addValidation, removeValidation } from 'vtex.address-form/helpers/'

import ShippingTable from './components/ShippingTable'
import getShippingEstimates from './queries/getShippingEstimates.gql'
import ShippingSimulatorLoader from './Loader'
import shippingSimulator from './shippingSimulator.css'
import { getNewAddress } from './utils'

const ShippingSimulator = ({
  intl,
  client,
  skuId,
  seller,
  country,
  styles,
}) => {
  const [address, setAddress] = useState(addValidation(getNewAddress(country)))
  const [shipping, setShipping] = useState({})
  const [loading, setLoading] = useState(false)
  const [isValid, setIsValid] = useState(false)

  const handleAddressChange = newAddress => {
    setAddress({
      ...address,
      ...newAddress,
    })
    const { postalCode } = address
    setIsValid(postalCode.valid)
  }

  const handleClick = e => {
    e.preventDefault()
    setLoading(true)
    const { postalCode } = removeValidation(address)
    client
      .query({
        query: getShippingEstimates,
        variables: {
          country,
          postalCode: postalCode,
          items: [
            {
              quantity: '1',
              id: skuId,
              seller,
            },
          ],
        },
      })
      .then(result => {
        setShipping(result.data.shipping)
      })
      .catch(error => {
        console.error(error)
      })
      .finally(() => {
        setLoading(false)
      })
  }

  if (!seller || !skuId) {
    return <ShippingSimulatorLoader {...styles} />
  }

  return (
    <Fragment>
      <div
        className={`${shippingSimulator.shippingContainer} t-small c-on-base`}
      >
        <AddressRules country={country} shouldUseIOFetching>
          <AddressContainer
            Input={StyleguideInput}
            address={address}
            onChangeAddress={handleAddressChange}
            autoCompletePostalCode
          >
            <PostalCodeGetter />
          </AddressContainer>
        </AddressRules>
        <Button
          onClick={handleClick}
          className={`${shippingSimulator.shippingCTA}`}
          disabled={!isValid}
          size="small"
          type="submit"
          block
          isLoading={loading}
        >
          {intl.formatMessage({ id: 'store/shipping.label' })}
        </Button>
      </div>
      <ShippingTable shipping={shipping} />
    </Fragment>
  )
}

ShippingSimulator.propTypes = {
  intl: intlShape.isRequired,
  client: PropTypes.object,
  skuId: PropTypes.string,
  seller: PropTypes.string,
  country: PropTypes.string.isRequired,
  /** Component and content loader styles */
  styles: PropTypes.object,
}

export default compose(
  withApollo,
  injectIntl
)(ShippingSimulator)
