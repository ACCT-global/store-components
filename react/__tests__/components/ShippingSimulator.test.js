import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import { renderWithIntl } from 'intl-helper'

import ShippingSimulator from '../../ShippingSimulator'

describe('<ShippingSimulator /> component', () => {
  const renderComponent = (customProps = {}) => {
    const props = {
      country: 'brazil',
      ...customProps,
    }
    const component = (
      <MockedProvider mocks={[]}>
        <ShippingSimulator {...props} />
      </MockedProvider>
    )

    return renderWithIntl(component)
  }

  it('should be able to mount and not break', () => {
    const component = renderComponent()
    expect(component).toBeTruthy()
  })

  it('should match snapshot without skuId and seller', () => {
    const component = renderComponent()
    expect(component).toMatchSnapshot()
  })

  it('should match snapshot with skuId and seller', () => {
    const props = {
      skuId: 'skuId',
      seller: 1,
    }
    const component = renderComponent(props)
    expect(component).toMatchSnapshot()
  })
})
