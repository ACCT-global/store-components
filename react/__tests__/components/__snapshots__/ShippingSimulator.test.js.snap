// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShippingSimulator /> component should match snapshot with skuId and seller 1`] = `
<form
  class="shippingContainer t-small c-on-base"
>
  <label
    class="shippingZipcodeLabel c-muted-2 db t-small mb3"
    for="shipping-zipcode"
  >
    Calculate shipping
  </label>
  <div
    class="flex"
  >
    <input
      id="shipping-zipcode"
      name="zipcode"
      type="text"
      value=""
    />
    <button
      type="button"
    >
       Ok 
    </button>
  </div>
</form>
`;

exports[`<ShippingSimulator /> component should match snapshot without skuId and seller 1`] = `
<div
  class="shippingContainer"
>
  <div
    class="content-loader-mock"
  >
    <rect
      height="100%"
      width="7em"
    />
    <rect
      height="100%"
      width="15em"
      x="8em"
    />
  </div>
</div>
`;
